pkg/jobconfig/files.go

```diff
@@ -610,34 +627,56 @@ func MakeRegexFilenameLabel(possibleRegex string) string {
 	return label
 }
 
-// IsGenerated returns true if the job was generated using prowgen
-func IsGenerated(job prowconfig.JobBase) bool {
-	_, generated := job.Labels[LabelGenerated]
-	return generated
+// IsGenerated returns true if the job was generated by the provided generator
+func IsGenerated(job prowconfig.JobBase, generator generator) (bool, error) {
+	requirement, err := labels.NewRequirement(LabelGenerator, selection.Equals, []string{string(generator)})
+	if err != nil {
+		return false, err
+	}
+	return labels.NewSelector().Add(*requirement).Matches(labels.Set(job.Labels)), nil
 }
 
-func isStale(job prowconfig.JobBase) bool {
-	genLabel, generated := job.Labels[LabelGenerated]
-	return generated && genLabel != string(NewlyGenerated)
+func isStale(job prowconfig.JobBase, generator generator, pruneLabels labels.Set) (bool, error) {
+	notNewlyGenerated, err := labels.NewRequirement(string(generator), selection.NotEquals, []string{string(NewlyGenerated)})
+	if err != nil {
+		return false, err
+	}
+	generatedByGenerator, err := labels.NewRequirement(LabelGenerator, selection.Equals, []string{string(generator)})
+	if err != nil {
+		return false, err
+	}
+	ls := labels.NewSelector().Add(*notNewlyGenerated).Add(*generatedByGenerator)
+	for label, value := range pruneLabels {
+		req, err := labels.NewRequirement(label, selection.Equals, []string{value})
+		if err != nil {
+			return false, err
+		}
+		ls = ls.Add(*req)
+	}
+	return ls.Matches(labels.Set(job.Labels)), nil
 }
 
-// Prune removes all prowgen-generated jobs that were not created by preceding
-// calls to GenerateJobs() (that method labels them as "newly generated"). All
-// remaining prowgen-generated jobs will be labeled as simply "generated" and
+// Prune removes all generated jobs of the supplied generator with values that are NOT newly-generated.
 // Prune() returns the resulting job config (which may even be completely empty).
-func Prune(jobConfig *prowconfig.JobConfig) *prowconfig.JobConfig {
+func Prune(jobConfig *prowconfig.JobConfig, generator generator, pruneLabels labels.Set) (*prowconfig.JobConfig, error) {
 	var pruned prowconfig.JobConfig
 
 	for repo, jobs := range jobConfig.PresubmitsStatic {
 		for _, job := range jobs {
-			if isStale(job.JobBase) {
+			stale, err := isStale(job.JobBase, generator, pruneLabels)
```

Comment:
This creates the same selector with the same input values for every job.

Create it once and use it here.

Reply:
Good call!